bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java

```diff
@@ -90,6 +91,7 @@
     private long lastUptime = 0;
     private long lastAlarmTs = 0;
     private long lastTimeoutErros = -1;
+    private final StopWatch watchdog = new StopWatch();
```

Comment:
You seem to be using this to implement some kind of timeout. I wouldn't suggest using a `StopWatch` to accomplish this. A watchdog is typically implemented as a scheduled or periodic task. It also seems like you are using the watchdog for multiple purposes, both as an update timeout and to track the time between updates. 

The watchdog should be single purpose otherwise you open yourself up to a lot of potential bugs.
Another issue is that your watchdog doesn't actually trigger anything on it's own, if the timeout expires nothing will happen until `refreshStatus` is called which unknown issues may prevent from happening in a timely manner. It might be better to implement your watchdog as a ScheduledFuture instead.

Reply:
> You seem to be using this to implement some kind of timeout. I wouldn't suggest using a `StopWatch` to accomplish this.  A watchdog is typically implemented as a scheduled or periodic task.

exactly, that's the reason why watchdog is a StopWatch. I could rename the variable if that is clearer for you.

>  It also seems like you are using the watchdog for multiple purposes, both as an update timeout and to track the time between updates. The watchdog should be single purpose otherwise you open yourself up to a lot of potential bugs.

I'm not ware of, where do you see this?

> Another issue is that your watchdog doesn't actually trigger anything on it's own, if the timeout expires nothing will happen until `refreshStatus` is called which unknown issues may prevent from happening in a timely manner. It might be better to implement your watchdog as a ScheduledFuture instead.

There are good reasons why I implemented that way
- I want 1 central point of dealing with device timeouts - that's refreshStatus()
- The device "has a last chance" to respond, this has a higher priority than setting it to OFFLINE, because of a timeout
- I don't want to have another periodic task to be synchronized - the watchdog affects the thing, that needs to be in sync
- The watchdog is important to detect dead devices, but not critical, therefore it's ok it the condition gets processed somewhat later

Even when thinking about it I still like the implementation. It's not invasive, users have confirmed that it works fine. As I said, I could rename the variable if this makes it clearer.

Reply:
Well if you already took all my points into consideration already then I'm fine with leaving it as is.